### DISCUSSION TITLE
Respect paths defined by installer plugins

### DISCUSSION THREAD

**Discussion by CDRO** [composer-unused/composer-unused#400] [Ideas]
In some cases, as in TYPO3, packages are not always installed in `vendor/<vendor>/<package>` as it would be expected by the plugin, in this case an error appears quickly and no scan is done.

To test this, simply execute this and run `composer-unused.phar` in the directory you just created:
```bash
composer create-project typo3/cms-base-distribution:^11 example-project-directory
```

As TYPO3 will probably shift all packages to the vendor folder in version 12 or 13 anyway, this might not be a problem in the future anymore, but regarding one of the the goals (helping remove packages before you upgrade, so you don't keep it around for nothing), this would probably be a game changer for the TYPO3 dev community, if it would work as expected 💪

**Comment by icanhazstring** [comment]
Hi @CDRO thanks for the report.
At the moment `composer-unused` is using the configuration `vendor-dir` if configured, otherwise falls back to default `vendor` as composer would do.

I will have to check the specific project for the structure and come back to update what needs to be done :)

**Comment by CDRO** [comment]
Thanks for considering, if I get the time to find the related information I'll try to help you with it and/or propose a patch to support this as well, this would probably benefit quite some projects as I don't think TYPO3 is the only framework or CMS having this kind of specifcs.

**Reply by icanhazstring** [reply]
Yes. There are still a lot of false positives for some frameworks still. Even for symfony which is most of the time config usage.

**Comment by icanhazstring** [comment]
I just did some investigation.
So the problem seems that typo3 is placing typo3 system extensions under `public/typo3/sysext`.

I don't know where normally the source is placed, maybe you can give me a hint here.

But I think the problem is placed here:
https://github.com/composer-unused/composer-unused/blob/main/src/Console/Command/UnusedCommand.php#L152
I assume the package to be installed inside the vendor dir, all the time. So it might be possible to use the `composer-unused.php` configuration to provide a list of additional directories for packages to be checked for source files, this way I don't need to adjust that much for finding required package sources.

So the only question remains where for typo3 the typical source is found.
This can then be also placed inside the configuration file and be checked like that:

```
return static function (Configuration $config): Configuration {
    return $config
        ->setAdditionalFilesFor('typo3/cms-base-distribution', [
            __FILE__,
            ...Glob::glob(__DIR__ . '/<additional directory>/*.php'),
        ]);
};
```

**Comment by CDRO** [comment]
Thanks for your quick investigation!

Usually, TYPO3 will put it's packages (aka extensions) as follows:
- system extensions are placed in `<web-dir>/typo3/sysext`
- userland extensions are placed in `<web-dir>/typo3conf/ext`
- everything else is placed in `vendor/` and is loaded as expected/as usual

The variable `<web-dir>` is set in `extra.typo3/cms.web-dir` (it defaults to `public/` if not defined differently):
```json
{
  "extra": {
    "typo3/cms": {
      "web-dir": "public/"
    }
  }
}
```

I guess Drupal has something similar, where the public modules are placed inside `web/`, but I'm not sure if that is configurable.

I've seen that symfony allows you to configure to have the public folder (as well as the `vendor` folder somewhere else, as well. (but you handle this case properly already):
https://symfony.com/doc/current/configuration/override_dir_structure.html#override-the-public-directory

I hope this helps!

**Reply by icanhazstring** [reply]
Ok got it. So the only problem remains that currently, you can't configure where packages are located.
If I add this feature into composer-unused it would be possible to use the configuration to adjust package and source dirs to your needs and everything should be working :)
